### PR TITLE
Add multiprocessing.sharedctypes stubs

### LINKED ISSUE #4047: Stub missing for multiprocessing.sharedctypes
`
from multiprocessing.sharedctypes import Synchronized
`

when mypy'ed produces this error: 

> playground\mpscerror.py:2: error: No library stub file for standard library module 'multiprocessing.sharedctypes'  [import]
    from multiprocessing.sharedctypes import Synchronized

### PR DESCRIPTION
**Pull request by hatal175**
Fixes #4047.

### CODE REVIEW

**Review by hauntsaninja — COMMENTED** [review]
Quick note in lieu of a full review: I think pytype now supports bound methods, so we should be able to replace a lot of the duplication in `multiprocessing.__init__`

**Comment by hatal175** [comment]
pyright gave me this issue:
https://github.com/hatal175/typeshed/runs/2377568497?check_suite_focus=true
which I could not overcome - that's why I basically copied all the functions to \_\_init__.

**Comment by hauntsaninja** [comment]
Thanks, yeah, this is the pytype issue: https://github.com/google/pytype/issues/761 Looks like the fix was only for bound methods within a file

**Comment by rchen152** [comment]
The fix in that pytype issue was *supposed* to be a complete one, not just for within a file. (I realize now that it looks confusing because I didn't actually add another comment to the issue, but you can see in the history that it was auto-closed by https://github.com/google/pytype/commit/0a50abf95a21b7d9211c5eb4805f5bc544c58b2e a couple months after the partial fix.)

Playing around with this PR, it looks like I missed a couple things - the aliased methods aren't found when they're from a base class, and there are some import resolution issues depending on the style of import used. I'll see if I can do a quick fix and open another pytype issue to track the problems if not.

**Comment by rchen152** [comment]
I wasn't able to figure out the import resolution issues, but I did get base class lookup working, so after the next pytype release, you should be able to do:
```
from multiprocessing import context
Array = context._default_context.Array
# etc.
```
Unfortunately, you have to use this *exact* style of import for pytype to resolve things correctly; `from multiprocessing.context import _default_context` or other variations don't work.

File: stdlib/multiprocessing/context.pyi [hunk 1]
```diff
@@ -53,22 +57,50 @@ class BaseContext(object):
...
-    def RawArray(self, typecode_or_type: Any, size_or_initializer: Union[int, Sequence[Any]]) -> Any: ...
-    # TODO: typecode_or_type param is a ctype with a base class of _SimpleCData or array.typecode Need to figure out
-    # how to handle the ctype
-    def Value(self, typecode_or_type: Any, *args: Any, lock: bool = ...) -> sharedctypes._Value: ...
-    # TODO: typecode_or_type param is a ctype with a base class of _SimpleCData or array.typecode Need to figure out
-    # how to handle the ctype
+    @overload
+    def RawValue(self, typecode_or_type: Type[_CT], *args: Any) -> _CT: ...
+    @overload
+    def RawValue(self, typecode_or_type: Union[str, Type[_CData]], *args: Any) -> Any: ...
```
Comment (srittau) [inline stdlib/multiprocessing/context.pyi:63]: `Type[_CData]` is already covered by the first overload and will never match in the second one. Same for `RawArraw` and the corresponding definitions in `.sharedctypes`.

File: stdlib/multiprocessing/sharedctypes.pyi [hunk 2]
```diff
@@ -1,43 +1,122 @@
...
+def synchronized(obj: ctypes.Array[c_char], lock: Optional[_LockLike] = ..., ctx: Optional[Any] = ...) -> SynchronizedString: ...
+@overload
+def synchronized(obj: ctypes.Array[_CT], lock: Optional[_LockLike] = ..., ctx: Optional[Any] = ...) -> SynchronizedArray[_CT]: ...
+@overload
+def synchronized(obj: _CT, lock: Optional[_LockLike] = ..., ctx: Optional[Any] = ...) -> SynchronizedBase[_CT]: ...
+
+class _AcquireFunc(Protocol):
+    def __call__(self, block: bool = ..., timeout: Optional[float] = ...) -> bool: ...
+
+class SynchronizedBase(ContextManager[bool], Generic[_CT]):
```
Comment (srittau) [inline stdlib/multiprocessing/sharedctypes.pyi:96]: We don't derive from `ContextManager` or other protocols, unless the implementation does. Please just copy `__enter__` and the cumbersome `__exit__` over.

File: stdlib/multiprocessing/sharedctypes.pyi [hunk 3]
```diff
@@ -1,43 +1,122 @@
...
+class SynchronizedArray(SynchronizedBase[ctypes.Array[_CT]], Generic[_CT]):
+    def __len__(self) -> int: ...
+    @overload
+    def __getitem__(self, i: int) -> _CT: ...
+    @overload
+    def __getitem__(self, s: slice) -> List[_CT]: ...
+    @overload
+    def __setitem__(self, i: int, o: _CT) -> None: ...
+    @overload
+    def __setitem__(self, s: slice, o: Iterable[_CT]) -> None: ...
```
Comment (srittau) [inline stdlib/multiprocessing/sharedctypes.pyi:116]: At least on Python 3.7 and 3.9, `SynchronizedArray` only allows the non-slice form.

File: stdlib/multiprocessing/sharedctypes.pyi [hunk 4]
```diff
@@ -1,43 +1,122 @@
...
+class SynchronizedBase(ContextManager[bool], Generic[_CT]):
+    acquire: _AcquireFunc = ...
+    release: Callable[[], None] = ...
+    def __init__(self, obj: Any, lock: Optional[_LockLike] = ..., ctx: Optional[Any] = ...) -> None: ...
+    def __reduce__(self) -> Tuple[Callable[..., Any], Tuple[Any, _LockLike]]: ...
+    def get_obj(self) -> _CT: ...
+    def get_lock(self) -> _LockLike: ...
+
+class Synchronized(SynchronizedBase[_SimpleCData[_T]], Generic[_T]):
+    value: _T = ...
```
Comment (srittau) [inline stdlib/multiprocessing/sharedctypes.pyi:105]: Nit: The `= ...` is redundant. (Also below with `SynchronizedString`.)

**Review by srittau — CHANGES_REQUESTED** [review]
Thank you. The mypy_primer failure can be ignored.

One general note: Please use built-in generics in Python-3-only stubs. E.g. `type[...]` instead of `Type[...]`. Also you can import other generics from `collections.abc` instead of `typing`. You don't need to change it for existing code (although you can), but something to keep in mind in the future.

More notes below.

Also please note that you should probably merge current master into this PR, update dependencies and re-run black as the formatting has changed in regards to overloads. Sorry about that!

File: stdlib/multiprocessing/context.pyi (hunk 1 above)
Comment (hatal175) [reply stdlib/multiprocessing/context.pyi:63]: _CData is covered but the union is not - unless I'm missing something. So If I'm giving this Union[str, _CArray] I want this to return Any. Right?

**Comment by hatal175** [comment]
I did the collections.abc change and the list/tuple change on the way in these files but I couldn't make the Type change - stubtest complained. I am working on python 3.8 so there might be a mypy bug there.

**Comment by srittau** [comment]
> I did the collections.abc change and the list/tuple change on the way in these files but I couldn't make the Type change - stubtest complained. I am working on python 3.8 so there might be a mypy bug there.
> Also note there isn't a single type[...] use in the repo.

There are unfortunately still a few mypy bugs around this.

File: stdlib/multiprocessing/context.pyi (hunk 1 above)
Comment (srittau) [reply stdlib/multiprocessing/context.pyi:64]: Type checkers should be able to build a union themselves. For example with mypy:

```python
from typing import overload, Union

@overload
def foo(x: str) -> str: ...
@overload
def foo(x: int) -> int: ...

x: Union[int, str]
reveal_type(foo(x))  # Revealed type is "Union[builtins.int, builtins.str]"
```

**Review by srittau — APPROVED** [review]

**Pull request merged**